### PR TITLE
Handle jobs execution order properly

### PR DESCRIPTION
**Pull request by firecow**

### CODE REVIEW

**Review by firecow — COMMENTED** [review]
Just review'ed everything here on github. LGTM

File: src/commander.ts [hunk 1]
```diff
@@ -3,126 +3,56 @@ import {Job} from "./job";
...
-            }
-        }
-
-        for (const job of jobs) {
-            await job.start(privileged);
+            jobPoolMap = new Map<string, Job>(jobs);
+            jobArgs.forEach(jobName => {
+                const job = Utils.getJobByName(jobs, jobName);
+                jobPoolMap.set(jobName, job);
+                potentialStarters = potentialStarters.concat(JobExecutor.getPastToWaitFor(jobs, stages, job));
```
Comment (moberghammer) [inline src/commander.ts:38]: Move if needs to this line only to depublicate

Comment (firecow) [reply src/commander.ts:38]: Job pool map is also special

**Review by moberghammer — APPROVED** [review]
Huge PR but looks okay as far as i can tell.

**Review by bcouetil — APPROVED** [review]
Tested  #110, it's working.

Too big for me to understand the code. Maybe you could explain the change in philosophy, for us to detect flaws in principle ?

**Comment by firecow** [comment]
@bcouetil Well, the pipeline and job execution was completely hacked together, and has been like that for a long time.
I've simplified job execution in the JobExecutor class, and added a lot of missing testes, to fix the issues reported.
So its not really a change in philosophy, as much as it is a move towards the current philosophy, with lots and lots of integration testing :smile:

**Comment by bcouetil** [comment]
OK, merge when you want :wink:

**Pull request merged**